Sources/Datadog/Core/System/AppStateListener.swift

```diff
@@ -0,0 +1,150 @@
+
+    /// Limits or extrapolates app state history to the given range
+    /// This is useful when you record between 0...3t but you are concerned of t...2t only
+    /// - Parameter range: if outside of `initialState` and `finalState`, it extrapolates; otherwise it limits
+    /// - Returns: a history instance spanning the given range
+    func take(between range: ClosedRange<Date>) -> AppStateHistory {
+        var taken = self
+        // move initial state to lowerBound
+        taken.initialState = Snapshot(
+            isActive: isActive(at: range.lowerBound),
+            date: range.lowerBound
+        )
+        // move final state to upperBound
+        taken.finalDate = range.upperBound
+        // filter changes outside of the range
+        taken.changes = taken.changes.filter { range.contains($0.date) }
+        return taken
+    }
+
+    var foregroundDuration: TimeInterval {
+        var duration: TimeInterval = 0.0
+        var lastActiveStartDate: Date?
+        let allEvents = [initialState] + changes + [finalState]
+        for event in allEvents {
+            if let startDate = lastActiveStartDate {
+                duration += event.date.timeIntervalSince(startDate)
+            }
+            if event.isActive {
+                lastActiveStartDate = event.date
+            } else {
+                lastActiveStartDate = nil
+            }
+        }
+        return duration
+    }
+
+    var didRunInBackground: Bool {
+        return !initialState.isActive || !finalState.isActive
+    }
+
+    private func isActive(at date: Date) -> Bool {
+        if date <= initialState.date {
+            // we assume there was no change before initial state
+            return initialState.isActive
+        } else if finalState.date <= date {
+            // and no change after final state
+            return finalState.isActive
+        }
+        var active = initialState.isActive
+        for change in changes {
+            if date < change.date {
+                break
+            }
+            active = change.isActive
+        }
+        return active
+    }
+}
+
+internal protocol AppStateListening: class {
+    var history: AppStateHistory { get }
+}
+
+internal class AppStateListener: AppStateListening {
+    typealias Snapshot = AppStateHistory.Snapshot
+
+    static var shared = AppStateListener()
+
+    private let dateProvider: DateProvider
+    private var _appStateHistory: AppStateHistory
+    var history: AppStateHistory {
+        objc_sync_enter(self)
+        defer { objc_sync_exit(self) }
+        _appStateHistory.finalDate = dateProvider.currentDate()
+        return _appStateHistory
+    }
+
+    private static var isAppActive: Bool {
+        if Thread.isMainThread {
+            return UIApplication.managedShared?.applicationState == .active
+        } else {
+            return DispatchQueue.main.sync {
+                UIApplication.managedShared?.applicationState == .active
+            }
+        }
```

Comment:
This if-check against the current thread looks odd. We can get rid of this if controlling the `AppStateListener` initialization and submitting this read asynchronously to the main thread.